### PR TITLE
refactor reducer

### PR DESCRIPTION
**Pull request by odeac**
## Overview
This refactoring is part of my learning process in the Reducer code, as I learn  how things work I refactor them such that I can read them better. 

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
RHOL-897

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [ ] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [ ] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [developer wiki](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain) for instructions.
- [ ] Your GitHub account is also an account with [Travis CI](https://travis-ci.org). Unit tests will not run on your PR unless you have an account with Travis. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 1]
```diff
@@ -195,20 +212,13 @@ object Reduce {
...
-            case e: EMethodBody =>
-              (for {
-                p <- evalExprToPar(Expr(e))
-                _ <- eval(p)(env, newRand, costAccountingAlg)
-              } yield ()).handleError(fTell.tell)
-            case _ => s.unit
-          }
+          exprHandler(expr, newRand)
         })
+        Parallel.parTraverse(par.sends.zipWithIndex.toList)(handle(mkTermHandler(evalExplicit), starts(0))),
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Please retain the original order of traverses this and the lines 217-221 should be above line 215

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -175,19 +175,6 @@ object Reduce {
         evalImpl(term, newRand)
       }
 
-      def handle2(start : Int)(texpr : (Expr, Int)) : M[Unit] = {
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Ideally, all your commit messages would start with a capital letter, and there would be no commits undoing / refactoring code introduced in the same PR (refactors would be applied in-place, in the commit introducing the refactored code). If you can do a rebase dropping this and the previous commit, I'll appreciate one :) Otherwise that's just a general remark for the future (mostly the commit message part).

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 3]
```diff
@@ -159,18 +159,17 @@ object Reduce {
...
 
         val newRand =
           if (starts(6) == 1)
             rand
           else if (starts(6) > 256)
-            rand.splitShort((start + idx).toShort)
+            rand.splitShort((starts(idx) + idx).toShort)
           else
-            rand.splitByte((start + idx).toByte)
+            rand.splitByte((starts(idx) + idx).toByte)
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: I'm not sure why this change is correct. Can you explain?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 4]
```diff
@@ -200,8 +192,10 @@ object Reduce {
           case _ => s.unit
         }
 
-      def mkJob[A](terms : Seq[A], handler : (A, Blake2b512Random) => M[Unit]) : M[List[Unit]] =
-        Parallel.parTraverse(terms.zipWithIndex.toList)(handle(handler))
+      def mkJob[A](terms: Seq[A], handler: (A, Blake2b512Random) => M[Unit]): M[List[Unit]] =
+        Parallel.parTraverse(terms.zipWithIndex.toList) {
+          case (term, idx) => handler(term, split(rand, idx))
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: can we make it a total funciton using `Function.tupled`?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 5]
```diff
@@ -192,18 +192,22 @@ object Reduce {
           case _ => s.unit
         }
 
-      def mkJob[A](terms: Seq[A], handler: (A, Blake2b512Random) => M[Unit]): M[List[Unit]] =
+      def mkJob[A](
+          terms: Seq[A],
+          handler: (A, Blake2b512Random) => M[Unit],
+          startIdx: Int
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: How about making `startIdx` the first parameter, for some nice alignment and faster-spotted copy-paste errors?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 6]
```diff
@@ -158,13 +158,13 @@ object Reduce {
...
-      def split(rand: Blake2b512Random, idx: Int): Blake2b512Random =
+      def split(rand: Blake2b512Random, startIdx: Int, idx: Int): Blake2b512Random =
         if (starts(6) == 1)
           rand
         else if (starts(6) > 256)
-          rand.splitShort((starts(idx) + idx).toShort)
+          rand.splitShort((starts(startIdx) + idx).toShort)
         else
-          rand.splitByte((starts(idx) + idx).toByte)
+          rand.splitByte((starts(startIdx) + idx).toByte)
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: turns out that change wasn't correct :D Again, Ideally you'd just fixup! + rebase

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 7]
```diff
@@ -192,23 +217,17 @@ object Reduce {
...
-        mkJob(par.bundles, mkTermHandler[Bundle](evalExplicit), 4),
-        mkJob(filteredExprs, exprHandler, 5)
-      ).parSequence.as(Unit)
+        EvalJob(par.sends, mkTermHandler[Send](evalExplicit)),
+        EvalJob(par.receives, mkTermHandler[Receive](evalExplicit)),
+        EvalJob(par.news, mkTermHandler[New](evalExplicit)),
+        EvalJob(par.matches, mkTermHandler[Match](evalExplicit)),
+        EvalJob(par.bundles, mkTermHandler[Bundle](evalExplicit)),
+        EvalJob(filteredExprs, exprHandler)
+      ).zipWithIndex
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Honeslty, I don't see any benefit from introducing the EvalJob. If you wanted to get rid of the hardcoded ints, I think you could've done:

```scala
        mkJob(par.sends, mkTermHandler[Send](evalExplicit), _),
        mkJob(par.receives, mkTermHandler[Receive](evalExplicit), _),
        mkJob(par.news, mkTermHandler[New](evalExplicit), _),
        mkJob(par.matches, mkTermHandler[Match](evalExplicit), _),
        mkJob(par.bundles, mkTermHandler[Bundle](evalExplicit), _),
        mkJob(filteredExprs, exprHandler, _)
      )
.zipWithIndex
.map(i, f -> f(i))
.parSequence.as(Unit)

```

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 8]
```diff
@@ -217,14 +211,18 @@ object Reduce {
...
         EvalJob(par.sends, mkTermHandler[Send](evalExplicit)),
         EvalJob(par.receives, mkTermHandler[Receive](evalExplicit)),
         EvalJob(par.news, mkTermHandler[New](evalExplicit)),
         EvalJob(par.matches, mkTermHandler[Match](evalExplicit)),
         EvalJob(par.bundles, mkTermHandler[Bundle](evalExplicit)),
         EvalJob(filteredExprs, exprHandler)
-      ).zipWithIndex
+      )
+
+      val starts = jobs.map(_.size).scanLeft(0)(_ + _).toVector
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: You've just convinced me the EvalJob has merit, due to it's size property :joy:

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 1 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: I don't understand what you mean here

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Here the intention was to make small steps such that each of them are obviously correct. This is why I introduced this function first and then it was obvious that only the exprHandler was different compared to the other one so I parameterized the difference. I think these small steps are a very good way to help the reviewer understand the refactoring process.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 3 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: It's not correct. I committed this without running the tests and I only realized it after I did the next step. Then I didn't undo because the fix was easy. See the commit "fix the tests"

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 4 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: This is function was abandoned during the refactoring but thanks for telling me about Function.tupled.  I didn't know about its existence.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 6 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: for some reason I got some rebase conflicts so it was easier to just add a new commit

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 5 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: outdated

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 8 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Good, because I was preparing the response for your previous comment :)

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 1 above)
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: This doesn't really matter because we traverse it in parallel.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 9]
```diff
@@ -125,6 +125,89 @@ object Reduce {
...
+          rand.splitShort((starts(startIdx) + idx).toShort)
+        else
+          rand.splitByte((starts(startIdx) + idx).toByte)
+
+      private def mkJob[A](
+          input: Seq[A],
+          handler: A => (Env[Par], Blake2b512Random, CostAccountingAlg[M]) => M[Unit]
+      ): EvalJob =
+        new EvalJob() {
+          override def run(starts: Vector[Int], startIdx: Int)(
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Previous implementation, when starting the evaluation of certain kind of terms didn't need to know about all start indexes (`starts`). Single `run` call on specific type of terms should only know about it's own index. It means that splitting needs to be done outside of `run`. Your `split` methods splits at `starts(startIdx) + idx` so it can be improved even more.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 10]
```diff
@@ -185,39 +224,13 @@ object Reduce {
...
-        }
-
       val jobs = List(
-        EvalJob(par.sends, mkTermHandler[Send](evalExplicit)),
-        EvalJob(par.receives, mkTermHandler[Receive](evalExplicit)),
-        EvalJob(par.news, mkTermHandler[New](evalExplicit)),
-        EvalJob(par.matches, mkTermHandler[Match](evalExplicit)),
-        EvalJob(par.bundles, mkTermHandler[Bundle](evalExplicit)),
-        EvalJob(filteredExprs, exprHandler)
+        EvalJob[Send](par.sends, evalExplicit),
```
[the remark this answers is not in the document]
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Tell that to the Scala compiler. Without putting the type explicitly, it doesn't compile

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 9 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: If you compare to the original implementation, the "rand splitting code" was referring start indices. Indeed, it might be a good idea to split the rngs upfront and zip them with the jobs

**Comment by odeac** [comment]
@goral09 By this refactoring a few things are achieved:

1. There are (almost) no magic numbers left in the code. Those are error prone and make it difficult to understand the intention of the code
2. The possibility of having subtle bugs because the following are not in sync: position in par, start index, position in the list of parTraverse's. Such connections make the code extremely brittle. 
3. It becomes obvious how the evaluation of expressions is different than the evaluation of the other terms. 
4. The rand splitting code is in one single function reused twice. This shows that we only have one rand splitting mechanism. This is important if we want to understand how the 
5. There is no more code duplication and the fact that the code is the same size it's just because those verbose signatures

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 11]
```diff
@@ -244,7 +239,12 @@ object Reduce {
       val starts = jobs.map(_.size).scanLeft(0)(_ + _).toVector
 
       jobs.zipWithIndex
-        .map { case (job, idx) => job.run(starts, idx)(env, rand, costAccountingAlg) }
+        .map {
+          case (job, idx) => {
+            val splitter = split(starts.last, starts(idx))(_, _)
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: If you already know `rand` here, why not make `splitter` of type `Int => Blake2b512Random` ?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Yes, and another advantage of making those small steps is you don't even have to show things that don't make it to the final shape, like these two commits: one introducing handle2 and the other one removing it. How you avoid showing such steps - and achieve the *ultimate* ease for the reviewer (less material to process) - is by doing a rebase.

Unless you mean that showing the introduction and retraction itself would help me understand sth. All I understood from those two is you changed your mind in the process.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 3 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Yeah, as I mentioned - that commit could've easily be a `fixup! [Original commit message]`, squashed automatically upon rebase (provided you have `git config --global rebase.autosquash true`)

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 11 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: good point

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 12]
```diff
@@ -125,6 +125,39 @@ object Reduce {
...
+      ): Blake2b512Random =
+        if (starts(6) == 1)
+          rand
+        else if (starts(6) > 256)
+          rand.splitShort((starts(startIdx) + idx).toShort)
+        else
+          rand.splitByte((starts(startIdx) + idx).toByte)
+
+      def apply[A](input: Seq[A], handler: (A, Blake2b512Random) => M[Unit]): EvalJob =
+        new EvalJob() {
```
[the remark this answers is not in the document]
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: ok
